modules/ROOT/pages/keycloak/retrieving-users-roles.adoc

```diff
@@ -1,52 +1,103 @@
 include::{partialsdir}/attributes.adoc[]
 
-= Retrieving a users roles
+= Access Control
+
+Once a `UserPrincipal` has been retrieved the link:http://www.keycloak.org/docs/latest/server_admin/index.html#roles[roles] of the user can be listed and checked. This can be used to perform client side access control, such as hiding UI components related to actions the user doesn't have permissions to perform.
+
+Roles are divided into two types. Resource roles which belong to the client the user has authenticated against, and Realm roles which belong to the realm the client is in.
 
 [role="primary"]
 .Android
 ****
-Once a `UserPrincipal` has been retrieved the link:http://www.keycloak.org/docs/latest/server_admin/index.html#roles[roles] of the user can be listed and checked. This can
-be used to perform client side access control, such as hiding UI components related to actions the
-user doesn't have permissions to perform.
-
-To list a users roles the `UserPrincipal#getRoles` method can be invoked.
-
-Roles are divided into two types. Resource roles which belong to the client the user has
-authenticated against, and Realm roles which belong to the realm the client is in.
-
-To list a users realm roles  the `UserPrincipal#getRealmRoles` method can be invoked and 
-to list a users resource roles the `UserPrincipal#getResourceRoles` can be invoked.
-
-In order to check if a user has a specific role you can invoke the `UserPrincipal#hasResourceRole`
-and `UserPrincipal#hasRealmRole` methods and provide the role name to check for.
-
+. Retrieve the authenticated user:
++
 [source,java]
 ----
 // authService already initialized.
 AuthService authService = MobileCore.getInstance().getService(AuthService.class);
 UserPrincipal currentUser = authService.currentUser();
+----
 
+. Use `hasRealmRole` and `hasResourceRole` to check for roles:
++
+[source,java]
+----
 boolean hasAdminPermissions = currentUser.hasRealmRole("user_admin");
-if (hasAdminPermissions) {
-    // Show some component.
-}
-
-// Check if a user has a role from a specific resource named my_resource.
 boolean isModerator = currentUser.hasResourceRole("my_resource", "user_moderator");
-if (isModerator) {
-    // Enable some button.
-}
 ----
+
 ****
 
 [role="secondary"]
 .iOS
 ****
+. Retrieve the authenticated user:
++
+[source,swift]
+----
+let currentUser = try AgsAuth.instance.currentUser()
+----
+
+. The user roles are available on the authenticated user:
++
+[source,swift]
+----
+let clientRoles = currentUser.clientRoles
+let realmRoles = currentUser.realmRoles
+----
+
+. Use `hasRealmRole` and `hasClientRole` to check for roles:
++
+[source,swift]
+----
+let hasAdminRole = currentUser.hasRealmRole("admin")
+let hasModeratorRole = currentUser.hasClientRole(client: "my_client", role: "moderator")
+----
 
 ****
 
 [role="secondary"]
 .Cordova
 ****
+Get the realm roles from the auth service:
+
+[source,javascript]
+----
+const realmRoles = authService.getRealmRoles();
```

Comment:
In Cordova, we miss some stuff. Created https://issues.jboss.org/browse/AEROGEAR-3350 for that.

But, we at least have `hasRealmRole` method.
I guess you didn't want to have that here?
https://github.com/aliok/mobile-docs/blob/a48f433aac7a784053a9d547e2c8c03e7c04ab60/modules/ROOT/pages/keycloak/retrieving-users-roles.adoc

Reply:
Yes, we miss some function in the js-sdk. I've covered this in the following note and suggest to use keycloak-js directly:

```
NOTE: No function to retrieve the resource roles is currently exported but you can use `extract` to get a handle to the underlying keycloak-js instance to retrieve the roles from there. Have a look at their documentation link:https://www.keycloak.org/docs/latest/securing_apps/index.html#_javascript_adapter[here].
```